File: 源代码/后端/src/main/java/cn/aicamera/backend/service/UserService.java
package cn.aicamera.backend.service;

import cn.aicamera.backend.dto.*;
import cn.aicamera.backend.exception.CustomException;
import cn.aicamera.backend.mapper.UserMapper;
import cn.aicamera.backend.model.User;
import cn.aicamera.backend.utils.JwtUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

@Service
public class UserService {
    @Autowired
    private UserMapper userMapper;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @Autowired
    private JwtUtil jwtUtil;

    @Autowired
    private TokenService tokenService;

    @Autowired
    private MinioService minioService;

    public void register(RegisterRequest request) {
        if (userMapper.findByEmail(request.getEmail()) != null) {
            throw new CustomException(200, "邮箱已存在");
        }
        User user = new User();
        user.setUsername(request.getUsername());
        user.setPassword(passwordEncoder.encode(request.getPassword()));
        user.setEmail(request.getEmail());
        userMapper.insert(user);
    }

    public String login(LoginRequest request) {
        User user = userMapper.findByEmail(request.getEmail());
//        if (user == null || !user.getPassword().equals(request.getPassword())) {
//            throw new CustomException(200, "用户名或密码错误");
//        }
        if (user == null || !passwordEncoder.matches(request.getPassword(), user.getPassword())) {
            throw new CustomException(200, "用户名或密码错误");
        }
        String token = jwtUtil.generateToken(user.getEmail());
        tokenService.storeToken(user.getEmail(), token);
        return token;
    }

    public void logout(String token) {
        if (token != null && token.startsWith("Bearer ")) {
            token = token.substring(7);
        }
        String username = jwtUtil.getUsernameFromToken(token);
        tokenService.removeToken(username);
    }

    public User selectUser(String token) {
        if (token != null && token.startsWith("Bearer ")) {
            token = token.substring(7);
        }
        String email = jwtUtil.getUsernameFromToken(token);
        User user = userMapper.findByEmail(email);
        return user;
    }
    public UserProfile updateUserProfile(String token, UserProfile request) {
        if (token != null && token.startsWith("Bearer ")) {
            token = token.substring(7);
        }
        String email = jwtUtil.getUsernameFromToken(token);
        User user = userMapper.findByEmail(email);
        user.setAge(request.getAge());
        user.setUsername(request.getUsername());
        user.setGender(request.getGender());
        user.setPreference(request.getPreference());
        userMapper.update(user);
        return request;
    }

    public String uploadAvatar(String token, MultipartFile avatar) {
        if (token != null && token.startsWith("Bearer ")) {
            token = token.substring(7);
        }

        String email = jwtUtil.getUsernameFromToken(token);
        User user = userMapper.findByEmail(email);

        if(!user.getAvatarUrl().isEmpty()) minioService.deleteFile(user.getAvatarUrl());
        String avatarUrl = minioService.uploadFile(avatar);

        user.setAvatarUrl(avatarUrl);
        userMapper.update(user);
        return avatarUrl;
    }

    public String validateLogin(String token) {
        // 去掉前缀
        if (token != null && token.startsWith("Bearer ")) {
            token = token.substring(7);
        }

        if (!jwtUtil.validateToken(token)) {
            throw new CustomException(200,"Token 无效或已过期");
        }

        // 更新 Token 的过期时间
        String username = jwtUtil.getUsernameFromToken(token);
        String newToken = jwtUtil.generateToken(username);
        tokenService.storeToken(username, newToken);
        return newToken;
    }
}

File: 源代码/前端/app/src/main/java/cn/aicamera/frontend/network/service/UserService.kt
package cn.aicamera.frontend.network.service

import cn.aicamera.frontend.model.GeneralResponse
import cn.aicamera.frontend.model.LoginRequest
import cn.aicamera.frontend.model.RegisterRequest
import cn.aicamera.frontend.model.SuccessResponse
import cn.aicamera.frontend.model.UserProfile
import okhttp3.MultipartBody
import okhttp3.ResponseBody
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.Header
import retrofit2.http.Multipart
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Part
import retrofit2.http.Query

interface UserService {
    // 用户注册
    @POST("/auth/register")
    suspend fun register(@Body request: RegisterRequest): Response<SuccessResponse>

    // 用户登录
    @POST("/auth/login")
    suspend fun login(@Body request: LoginRequest): Response<GeneralResponse<String>>

    // 退出登录
    @GET("/auth/logout")
    suspend fun logout(): Response<SuccessResponse>

    // 检查用户登录信息
    @GET("/auth/validate-login")
    suspend fun validateLogin(): Response<GeneralResponse<String>>

    // 获取用户信息
    @GET("/user/info")
    suspend fun getUserProfile(): Response<GeneralResponse<UserProfile>>

    @GET("/user/avatar")
    suspend fun getAvatar(@Query("avatarUrl") avatarUrl: String): Response<ResponseBody>

    // 更新用户信息，返回新用户信息
    @PUT("/user/update")
    suspend fun updateUserProfile(@Body request: UserProfile): Response<GeneralResponse<UserProfile>>

    // 上传头像，返回头像链接
    @Multipart
    @POST("/user/upload-avatar")
    suspend fun uploadAvatar(@Part avatar: MultipartBody.Part): Response<GeneralResponse<String>>
}